Introduce OAuth2 Step Up Authentication for Bearer token authentication (#47339)
quarkusio/quarkus: docs/src/main/asciidoc/security-oidc-bearer-token-authentication.adoc

For more complex setup involving multiple tenants please see the xref:security-openid-connect-multitenancy.adoc#programmatic-startup[Programmatic OIDC start-up for multitenant application]
section of the OpenID Connect Multi-Tenancy guide.

> == OAuth 2.0 Step Up Authentication

Reviewer: ```suggestion
== Step Up Authentication
```

Just a minor suggestion as we also refer to DPoP and other OAuth2 specs without `OAuth2` in the section title, given the doc is about OIDC :-)

The `io.quarkus.oidc.AuthenticationContext` annotation can be used to enforce required authentication level for the Jakarta REST resource classes and methods.

> The https://datatracker.ietf.org/doc/rfc9470/[OAuth 2.0 Step Up Authentication Challenge Protocol] introduced a mechanism for resource servers like Quarkus to challenge authentication requirements.

Reviewer: ```suggestion
The https://datatracker.ietf.org/doc/rfc9470/[OAuth 2.0 Step Up Authentication Challenge Protocol] introduces a mechanism for resource servers to request stronger token authentication methods.
```
May be something like that...
And change the sentence order, this one goes first and an info about the annotation - second

> Consider following example:

Reviewer: ```suggestion
Consider the following example:
```

> [source,properties]

Reviewer: IMHO it can be added after the main introduction showing the annotation

----
quarkus.http.auth.proactive=false <1>
----
<1> Disable proactive authentication so that the `@AuthenticationContext` annotation can be matched with the endpoint before Quarkus authenticates incoming requests.

[source,java]
----
package io.quarkus.it.oidc;

import io.quarkus.oidc.AuthenticationContext;
import jakarta.ws.rs.GET;
import jakarta.ws.rs.Path;

@Path("hello")
public class HelloResource {

    @AuthenticationContext("myACR") <1>
    @GET
    public String hello() {
        return "hello";
    }

}
----

> <1> Only HTTP requests with valid Bearer access token with `acr` claim value that contains `myACR` can access this endpoint.

Reviewer: ```suggestion
<1> Bearer access token's  `acr` claim must include the `myACR`  authentication context class.
```

If the Bearer token claim `acr` does not contain `myACR`, Quarkus returns the Authentication Requirements Challenge indicating required `acr_values`:

----
HTTP/1.1 401 Unauthorized
WWW-Authenticate: Bearer error="insufficient_user_authentication",
 error_description="A different authentication level is required",
 acr_values="myACR"
----

> A client receiving the challenge from the Quarkus server carrying the `insufficient_user_authentication` error code should parse `acr_values` from the `WWW-Authenticate` header and use them when constructing new authorization request.

Reviewer: ```suggestion
When a client such as Single-page application (SPA) receives a challenge with the `insufficient_user_authentication` error code, it must parse `acr_values`, request a new user login which must meet the `act_values` constraints, and use a new access token to access Quarkus.
```

Not sure, but may be it can be more informative ?

[NOTE]
====

> The `io.quarkus.oidc.AuthenticationContext` annotation can be used to enforce required authentication level for a WebSockets Next server endpoint.

Reviewer: ```suggestion
The `io.quarkus.oidc.AuthenticationContext` annotation can also be used to enforce required authentication level for a WebSockets Next server endpoint.
```
